File: binary-trees/find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree/find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree.java
class Solution {

  public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
      this.val = val;
    }
  }

  public final TreeNode getTargetCopy(final TreeNode original, final TreeNode cloned, final TreeNode target) {
    return dfs(cloned, target.val);
  }

  public TreeNode dfs(TreeNode root, int target) {
    if (root == null)
      return null;

    if (root.val == target) {
      return root;
    }

    TreeNode left = dfs(root.left, target);
    if (left != null)
      return left;
    TreeNode right = dfs(root.right, target);
    if (right != null)
      return right;

    return null;
  }
}

File: binary-trees/find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree/find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree.ts
class TreeNode {
  val: number;
  left: TreeNode | null = null;
  right: TreeNode | null = null;

  constructor(val: number) {
    this.val = val;
  }
}

function getTargetCopy(
  original: TreeNode | null,
  cloned: TreeNode | null,
  target: TreeNode | null
): TreeNode | null {
  return dfs(cloned, target);
}

function dfs(root: TreeNode | null, target: TreeNode | null): TreeNode | null {
  if (!root) return null;

  if (root.val === target?.val) return root;

  const left = dfs(root.left, target);
  if (left) return left;

  const right = dfs(root.right, target);
  if (right) return right;

  return null;
}
